### PR TITLE
Add new analyses and bring code more in line with JASP standards

### PR DESCRIPTION
**Pull request by jansim**
This PR jointly adds the new analyses / features from [LSLindeloo/jaspPower/powerModule](https://github.com/LSLindeloo/jaspPower/tree/powerModule) by @LSLindeloo and the rewrtie of large parts of the codebase to be more in line with JASP standards [jansim/jaspPower/#9](https://github.com/jansim/jaspPower/pull/9).

The JASP-style rewrite mainly came with a transition from an R6 based object oriented style to a more functional style, following JASP function naming conventions e.g. prefixing non-exported functions with a `.`.

This PR also includes an upgrade and matching fixes to use the new unit test specification.

@LSLindeloo not sure whether you may want to elaborate on the new analyses / features here?

### CODE REVIEW

File: R/pwr_1p.R
Comment (Kucharssim) [inline R/pwr_1p.R:450]: html tahs and unicode symbols should be passed into gettextf as arguments; I assume we don't want the translators with changing e.g., `<i>` to `<br>`!

File: R/pwr_1p.R
Comment (Kucharssim) [inline R/pwr_1p.R:486]: for these checks you could use `jaspBase::isTryError`

File: R/pwr_1var.R
Comment (Kucharssim) [inline R/pwr_1var.R:10]: This seems to be on top of each of the analyses, could you move this into a function and just reuse the function?

File: R/pwr_2p.R
Comment (Kucharssim) [inline R/pwr_2p.R:26]: These parts are also a bit repetetive...

For example, instead of having in each analysis

```r
if(options[["text"]])
  .populateIntro(jaspResults, options)
```

You could just define `.populateIntro` with
```
.populateIntro <- function(jaspResults, options) {
  if(!options[["text"]]) return()

  # rest of the original function
}
```

Then you don't have to do the check every time you call this function and as a result the analysis functions become a bit more readable.

**Review by Kucharssim — CHANGES_REQUESTED** [review]
Looks very nice. I have not tested very thoroughly in JASP, but it seems to work ok for majority of "reasonable" inputs. Pushing it to less sensible regions leads to informative error messages, which is good.

I provided some code comments regarding some style and code duplication - these would be nice to polish but it's not absolutely necessary. There are some issues with the use of `gettextf` which we definitely need to fix before merging though.

File: inst/qml/Power.qml
Comment (shun2wang) [inline inst/qml/Power.qml:83]: Hi, we shoud not mark a empty string as to be translated:-) and [here](https://github.com/jasp-stats/jaspModuleTemplate/pull/8) we will have a test to check this.

File: R/commonPower.R
Comment (jansim) [reply R/commonPower.R:15]: I think the max value is already set in Power.qml:159 , so this is just an additional check. I think it makes sense to leave it in, given that there's also the idea floating around of supporting the use of jasp modules within R (or am I mistaken here? 🤔)

File: R/commonPower.R
Comment (jansim) [reply R/commonPower.R:73]: I think I cought every one of these now, thanks!

File: R/commonPower.R
Comment (jansim) [reply R/commonPower.R:72]: Good one, I've always been a bit hesitant about using `switch` in R due to its weird implementation as a function. This is a really good use case of it though so will change it, ty! 👍️

File: R/commonPowerPlotting.R
Comment (jansim) [reply R/commonPowerPlotting.R:100]: Good catch! ✅

File: R/pwr_2p.R
Comment (jansim) [reply R/pwr_2p.R:26]: Ah, yeah, this is the part where I struggled a bit to find a nice way of porting this over from the object oriented programming style (which solves this more elegantly).

Will update this for any common functions, I think for the analysis specific functions it's probably more readable this way? And it would be duplicated either way 🤔

File: R/commonPower.R
Comment (jansim) [reply R/commonPower.R:119]: This is actually coming from jpower, not 100% sure what Richard changed here and whether it's worth digging into? I updated the comment for now to clearly reference the source of the code 👍️

**Comment by jansim** [comment]
Thank you for these very detailed and super helpful tips!! 🙏🙏

I did a first pass on this, adressing most of the comments. Feel free to re-open comments, if you disagree with me marking it as resolved (this was also largely for my own tracking of what's addressed).

I managed to deal with about 50% of the `gettextf` related issues, unfortunately these take a huge amount of time since they are scattered throughout the whole codebase and can't be solved with find / replace (at least that I know of). @LSLindeloo do you think you could maybe help here in fixing the last ones of these? This is mainly about getting rid of any HTML in translated strings and replacing unordered `%s` with `%1$s`, `%2%s`, ... for strings with multiple parameters. Either way it's probably good to have a second pair of eyes on this since it's a lot of just shuffling things around.

File: R/commonPower.R
Comment (Kucharssim) [reply R/commonPower.R:15]: > supporting the use of jasp modules within R (or am I mistaken here? 🤔)

No you are correct, but the idea is that if the user calls a JASP function in R, the options are sent to QML for validation first; the input in R will go through the same checks as in the JASP GUI - to exactly avoid having to implement these checks both in R and QML.

File: R/commonPower.R
Comment (Kucharssim) [reply R/commonPower.R:119]: ah ok, then this is ok

File: R/commonPowerPlotting.R
Comment (Kucharssim) [reply R/commonPowerPlotting.R:4]: > I think themeJasp itself actually deprecated? 

You are right!

In this case I would compare to

```r
plotObject +
  jaspGraphs::geom_rangeframe() +
  jaspGraphs::themeJaspRaw(legend.position = "right")
```

The idea is to have a uniform style of plots across JASP, so it makes sense to stick to that. Unless there is a reason why the default JASP style does not work for the particular plot (that does not seem the case here but you can convince me otherwise).

> If this is important, I guess we could just quickly change it over as this is handled in the function anyways. 🤷‍♂️

Yes, it's nice that you wrapped it in a separate function!

An alternative is to let the user switch between different themes (this is for example done in Mixed Models module) if you think it's worth for people to try different looks. But usually it's not worth the effort.

**Comment by jansim** [comment]
@Kucharssim I've updated the plots to use the jaspGraphs theme, however, this requires the snapshots to be updated and I can only do so for macOS (since some of the plots are platform dependent). Could you re-run the tests locally and upload the plots if you are based on either Windows or Linux? (Or do you know someone from the team who could do this?)

**Comment by jansim** [comment]
Thank you, Don! No idea why it wasn't possible, but I merged the changes and all tests pass :tada:

File: inst/qml/Power.qml
Comment (Kucharssim) [inline inst/qml/Power.qml:170]: This option needs `max`

**Review by Kucharssim — CHANGES_REQUESTED** [review]
There are still some details to fix in the QML validation, otherwise there is only some gettexting to be done; I would recommend one more walk through the code to polish that out. Otherwise it looks great!

> I also tried to catch the last remaining occurences of HTML in gettext / gettextf, not sure whether there's an option of exporting all the strings to catch any remaining bits that might've slipped through?

Well we have a GitHub action that will generate the to-be-translated strings. But usually that runs after we merge a PR. Of course we could merge it as is and make another PR to clean up later, but it would be really better to do as much as we can now already. Unless @shun2wang or @vandenman have any other idea how to make this process more convenient.

> @jansim I've opened https://github.com/jansim/jaspPower-fork-only/pull/1 which regenerates the test figures. For some reason I can't push to this PR directly, otherwise I would have done so.

It seems I was the only administrator of this repo; I just changed the permissions to jasp teams according to other modules.

**Comment by shun2wang** [comment]
> I also tried to catch the last remaining occurences of HTML in `gettext` / `gettextf`, not sure whether there's an option of exporting all the strings to catch any remaining bits that might've slipped through?

Hi, [here](https://github.com/jasp-stats/jasp-actions/blob/dd3afbaa5162c1efd04f58db6c6e06639585bde5/translations/i18nCheck.R#L10) we have a action to catch all the strings to be translated as a dataframe. we didn't export them in the GitHub wrokflow. but we can do this if necessary.
You can also run this tool locally:

```r
library(potools)
rPotData <- potools::get_message_data(dir = ".", verbose = TRUE)
View(rPotData )
```
Then you can see all the strings with translations listed in the call column (this also contains additional` stop/warning...` calls, usually you should ignore them because JASP now only treats calls explicitly marked as `gettext/gettextf` as strings to be translated).

File: R/commonPower.R
Comment (jansim) [reply R/commonPower.R:15]: Oh wow, what a fancy solution! At some point I'll have to take a look how you parsed the QML and all! 👀 

I'll remove the line then 👍

**Comment by jansim** [comment]
Thank you for the detailed feedback again and that amazing snippet (@shun2wang)!

I've eradicated the last bits of HTML in translated strings and updated the QML to use the interval setting correctly 👍️

I used the following snippet to detect all messages containing HTML:

```R
library(potools)
library(tidyverse)
potools::get_message_data(dir = ".", verbose = TRUE) %>% 
  filter(
    str_detect(msgid, fixed("<b>")) |
    str_detect(msgid, fixed("<i>")) |
    str_detect(msgid, fixed("<p>"))
  ) %>% 
  View()
```

File: R/pwr_2var.R
Comment (shun2wang) [inline R/pwr_2var.R:284]: These seem also not need to be translated, because translator will do nothing for them. but it's ok if you don't have time to clean these.

**Review by shun2wang — APPROVED** [review]

File: R/pwr_2var.R
Comment (jansim) [reply R/pwr_2var.R]: Good catch! I replaced gettextf with sprintf in these cases (couldn't be bothered to change it into paste, but maybe @LSLindeloo can update parts here if the code around it is being changed anyways at some point, AFAIK he did want to do some minor changes there)

Comment (jansim) [reply R/pwr_2var.R]: Also FYI if anyone runs into a similar problem in the future, I was able to easiliy find these strings by searching for the ones with a high ratio of `%` compared to overall text length 😄

**Review by Kucharssim — APPROVED** [review]

**Comment by Kucharssim** [comment]
Looks great! Thanks @jansim for the additional effort you had to put in and @shun2wang for a another pair of eyes in the review!

**Pull request merged**
